crypto/evp/keymgmt_lib.c

```diff
@@ -0,0 +1,289 @@
+        *q = *p;
+        if (p->key == NULL)
+            break;
+
+        q->data = NULL;          /* In case the provider used it */
+        q->return_size = 0;
+    }
+
+    return params;
+}
+
+typedef union align_block_un {
+    OSSL_UNION_ALIGN;
+} ALIGN_BLOCK;
+
+#define ALIGN_SIZE  sizeof(ALIGN_BLOCK)
+
+static void *allocate_params_space(OSSL_PARAM *params)
+{
+    unsigned char *data = NULL;
+    size_t space;
+    OSSL_PARAM *p;
+
+    for (space = 0, p = params; p->key != NULL; p++)
+        space += ((p->return_size + ALIGN_SIZE - 1) / ALIGN_SIZE) * ALIGN_SIZE;
+
+    data = OPENSSL_zalloc(space);
+
+    for (space = 0, p = params; p->key != NULL; p++) {
+        p->data = data + space;
+        space += ((p->return_size + ALIGN_SIZE - 1) / ALIGN_SIZE) * ALIGN_SIZE;
+    }
+
+    return data;
+}
+
+void *evp_keymgmt_export_to_provider(EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
+{
+    size_t i;
+
+    /*
+     * If there is an underlying legacy key and it has changed, invalidate
+     * the cache of provider keys.
+     */
+    if (pk->pkey.ptr != NULL) {
+        /*
+         * If there is no dirty checker, this key can't be used
+         * with providers
+         */
+        if (pk->ameth->is_dirty == NULL)
+            return NULL;
+
+        if (pk->ameth->is_dirty(pk))
+            evp_keymgmt_clear_pkey_cache(pk);
+    }
+
+    /*
+     * See if we have exported to this provider already.
+     * If we have, return immediately.
+     */
+    for (i = 0;
+         i < OSSL_NELEM(pk->pkeys) && pk->pkeys[i].keymgmt != NULL;
+         i++) {
+        if (keymgmt == pk->pkeys[i].keymgmt)
+            return pk->pkeys[i].provkey;
+    }
+
+    /*
+     * If there are no cached provider keys, try to export a legacy key
+     * to the provider.
+     */
+    if (i == 0) {
+        /* If there is no legacy key, give up */
+        if (pk->pkey.ptr == NULL)
+            return NULL;
+
+        /* Also, if there is no export function, give up */
+        if (pk->ameth->export_to == NULL)
+            return NULL;
+
+        /* Otherwise, simply use it and return the result */
+        EVP_KEYMGMT_up_ref(keymgmt);
+        pk->pkeys[0].keymgmt = keymgmt;
+        pk->pkeys[0].provkey = pk->ameth->export_to(pk, keymgmt);
+        return pk->pkeys[0].provkey;
+    }
+
+    /*
+     * There are cached provider keys, so import from the first one
+     * and export the imported data to the new provider.
+     * If it's impossible to import from the cached provider, give up
+     * (transfering keys between providers can't be supported in that
+     * case)
+     */
+    if (keymgmt->importkey_priv != NULL
+        && pk->pkeys[0].keymgmt->exportkey_priv != NULL)
+    {
+        const OSSL_PARAM *paramdefs = NULL;
+        OSSL_PARAM *params = NULL;
+        void *data = NULL;
+        void *provctx = ossl_provider_ctx(EVP_KEYMGMT_provider(keymgmt));
+
+        paramdefs = pk->pkeys[0].keymgmt->exportkey_priv_types();
```

Comment:
We seem to be talking specifically about private keys here. What about public keys?

Reply:
Ah, good question.  The "traditional" OpenSSL setup is that the private key *always* also holds the public key, while the public key doesn't hold the private key.  Considering we also usually pass the numbers that were used to calculate the keys themselves, it's really rather pointless *not* to include the public bits with the private bits, but maybe this is something we'd want to rethink?

Otherwise, maybe we should have "key" and "key_pub" rather than "key_priv" and "key_pub".

I'm open for alternatives